### PR TITLE
Google Ads Audience - add deletion support

### PR DESCRIPTION
**Pull request by akashrpo**
## Description of the change

> add deletion support for Google Ads Audience

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Comment by utsabc** [comment]
@akashrpo GA generally refers to Google Analytics, can we rephrase this to Google Ads Audience

File: v0/destinations/google_adwords_remarketing_lists/util.js [hunk 1]
```diff
@@ -114,20 +114,44 @@ const gaAudienceProxyRequest = async request => {
...
+
+  // If two operations are needed i.e create and remove both
+  if (Object.keys(request.body.JSON).length > 1) {
+    response = await gaAudienceProxyRequest(
+      request,
+      Object.values(request.body.JSON)[1]
+    );
+  }
+
+  return response;
```
Comment (ItsSudip) [inline v0/destinations/google_adwords_remarketing_lists/util.js:152]: This logic doesn't look right to me. If some user sends both add and remove calls but due to some mistakes in attributes, the add call failed. But the remove call doesn't have any mistakes though we are discarding both the calls. If we can do separate calls for add and remove that would be ideal. In facebook custom audience, we support this. CC @utsabc

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 2]
```diff
@@ -133,62 +133,45 @@ const populateIdentifiers = (attributeArray, { Config }) => {
...
-    outputPayload.operations.push(operations);
+    const outputPayload = constructPayload(message, offlineDataJobsMapping);
+    outputPayload.operations = [];
+    // breaking the userIdentiFier array in chunks of 20
+    const userIdentifierChunks = returnArrayOfSubarrays(
+      userIdentifiersList,
+      20
+    );
+    // putting each chunk in different create/remove operations
+    if (key === "add") {
```
Comment (utsabc) [inline v0/destinations/google_adwords_remarketing_lists/transform.js:151]: Here all the `userIdentifiers` will go to `create` operation and `remove` as we are not mapping which `identifiers` are part of `create` and which are part of `remove`

Example
Input
```
{
    "userId": "user1",
    "anonymousId": "anon-id-new",
    "event": "event1",
    "type": "audienceList",
    "properties": {
        "listData": {
            "add": [
                {
                    "email": "alex@example.com",
                    "phone": "1234567890"
                },
                {
                    "email": "emily@example.com",
                    "phone": "1234567890"
                }
            ],
            "remove": [
                {
                    "email": "alex@example.com",
                    "phone": "1234567890"
                },
                {
                    "email": "emily@example.com",
                    "phone": "1234567890"
                }
            ]
        },
        "enablePartialFailure": true
    },
    "context": {
        "ip": "1.2.3.40",
        "library": {
            "name": "http"
        }
    },
    "timestamp": "2020-02-02T00:23:09.544Z"
}
```

Output
```
{
  create: {
    enablePartialFailure: true,
    operations: [
      {
        create: {
          userIdentifiers: [
            {
              hashedEmail: "6db61e6dcbcf2390e4a46af426f26a133a3bee45021422fc7ae86e9136f14110",
            },
            {
              hashedPhoneNumber: "c775e7b757ede630cd0aa1113bd102661ab38829ca52a6422ab782862f268646",
            },
            {
              hashedEmail: "6599ce4d8f9d3289820e0af8c795c5a06bdb512fe9cd3baaaeb132d35e641b09",
            },
            {
              hashedPhoneNumber: "c775e7b757ede630cd0aa1113bd102661ab38829ca52a6422ab782862f268646",
            },
          ],
        },
      },
    ],
  },
  remove: {
    enablePartialFailure: true,
    operations: [
      {
        remove: {
          userIdentifiers: [
            {
              hashedEmail: "6db61e6dcbcf2390e4a46af426f26a133a3bee45021422fc7ae86e9136f14110",
            },
            {
              hashedPhoneNumber: "c775e7b757ede630cd0aa1113bd102661ab38829ca52a6422ab782862f268646",
            },
            {
              hashedEmail: "6599ce4d8f9d3289820e0af8c795c5a06bdb512fe9cd3baaaeb132d35e641b09",
            },
            {
              hashedPhoneNumber: "c775e7b757ede630cd0aa1113bd102661ab38829ca52a6422ab782862f268646",
            },
          ],
        },
      },
    ],
  },
}
```

File: v0/destinations/google_adwords_remarketing_lists/util.js (hunk 1 above)
Comment (utsabc) [reply v0/destinations/google_adwords_remarketing_lists/util.js:152]: As discussed we will handle this at transformer and keep things clean at Proxy

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 3]
```diff
@@ -133,62 +133,45 @@ const populateIdentifiers = (attributeArray, { Config }) => {
 const createPayload = (message, destination) => {
   const { listData } = message.properties;
 
-  if (!listData.add || !isArray(listData.add)) {
-    throw new CustomError(
-      "[Google_adwords_remarketing_list]::add is not present inside listData. Aborting message.",
-      400
+  let outputPayloads = {};
+  const typeOfOperation = Object.keys(listData);
```
[the remark this answers is not in the document]
Comment (utsabc) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:137]: We can make things cleaner here. Also `typeOfOperation` should be validated against an enum. And we can use switch case to make things clean instead of if else

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 4]
```diff
@@ -133,62 +133,45 @@ const populateIdentifiers = (attributeArray, { Config }) => {
...
+  typeOfOperation.forEach(key => {
+    const userIdentifiersList = populateIdentifiers(
+      listData[`${key}`],
```
[the remark this answers is not in the document]
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js]: Done

File: v0/destinations/google_adwords_remarketing_lists/transform.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:138]: used switch case

File: v0/destinations/google_adwords_remarketing_lists/util.js (hunk 1 above)
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/util.js]: modified to handle at transformer itself

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 5]
```diff
@@ -136,45 +136,50 @@ const createPayload = (message, destination) => {
...
-      // for remove operation
-      userIdentifierChunks.forEach(element => {
-        const operations = {
-          remove: {}
-        };
-        operations.remove.userIdentifiers = element;
-        outputPayload.operations.push(operations);
-      });
-      outputPayloads = { ...outputPayloads, remove: outputPayload };
+      logger.log(`listData's ${key} is not valid.`);
```
Comment (utsabc) [inline v0/destinations/google_adwords_remarketing_lists/transform.js:179]: > used switch case
Its still if else

Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:179]: This is outdated one

File: v0/destinations/google_adwords_remarketing_lists/util.js [hunk 6]
```diff
@@ -100,9 +99,10 @@ const gaAudienceProxyRequest = async request => {
   );
   if (
     !firstResponse.success &&
-    !isHttpStatusSuccess(firstResponse.response.status)
-  )
+    !isHttpStatusSuccess(firstResponse.response.response.status)
```
Comment (utsabc) [inline v0/destinations/google_adwords_remarketing_lists/util.js:102]: Is this correct how was this working earlier?
@ItsSudip @akashrpo

File: v0/destinations/google_adwords_remarketing_lists/util.js [hunk 7]
```diff
@@ -114,14 +114,15 @@ const gaAudienceProxyRequest = async request => {
     headers,
     method,
     jobId,
-    body
+    request.body.JSON
```
Comment (utsabc) [inline v0/destinations/google_adwords_remarketing_lists/util.js:117]: Let's use Object destructuring at top to get this much safer

File: v0/destinations/google_adwords_remarketing_lists/transform.js (hunk 5 above)
Comment (utsabc) [reply v0/destinations/google_adwords_remarketing_lists/transform.js]: Resolving this convo

File: v0/destinations/google_adwords_remarketing_lists/util.js (hunk 6 above)
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/util.js:102]: It was not working earlier. Code coverage was not here

Comment (utsabc) [reply v0/destinations/google_adwords_remarketing_lists/util.js:102]: I mean didn't we do negative test cases then. Actual integration tests to destination with incorrect parameters to mimick these failure case. @ItsSudip ?

File: v0/destinations/google_adwords_remarketing_lists/util.js (hunk 7 above)
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/util.js]: refactored

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 8]
```diff
@@ -132,66 +132,59 @@ const populateIdentifiers = (attributeArray, { Config }) => {
...
+            };
+            operations.remove.userIdentifiers = element;
+            outputPayload.operations.push(operations);
+          });
+          outputPayloads = { ...outputPayloads, remove: outputPayload };
+          break;
+        default:
+      }
+    } else {
+      logger.log(`listData's ${key} is not valid.`);
```
Comment (ItsSudip) [inline v0/destinations/google_adwords_remarketing_lists/transform.js:179]: Can you update this message with something like, "we are supporting add and remove only ${key} is not supported."

File: v0/destinations/google_adwords_remarketing_lists/util.js (hunk 6 above)
Comment (ItsSudip) [reply v0/destinations/google_adwords_remarketing_lists/util.js:102]: ```isHttpStatusSuccess``` is getting undefined and returning false, so it was working.

File: v0/destinations/google_adwords_remarketing_lists/transform.js (hunk 8 above)
Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js]: Modified

**Comment by utsabc** [comment]
Take pull from master for `skipVerify` option

**Comment by utsabc** [comment]
Take pull from master for `skipVerify` option

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

File: v0/destinations/google_adwords_remarketing_lists/transform.js [hunk 9]
```diff
@@ -132,66 +135,61 @@ const populateIdentifiers = (attributeArray, { Config }) => {
...
-  //     operations[`${key}`].userIdentifiers = element;
-  //     outputPayload.operations.push(operations);
-  //   });
-  //   outputPayloads.push(outputPayload);
-  // });
-  // return outputPayloads;
+  return outputPayloads;
 };
 
 const processEvent = async (metadata, message, destination) => {
```
Comment (saikumarrs) [inline v0/destinations/google_adwords_remarketing_lists/transform.js:191]: Do we have test cases to cover all the validations performed here?

Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:191]: have added all the test cases.

Comment (saikumarrs) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:191]: Can you please point me to each of those TCs?

Comment (akashrpo) [reply v0/destinations/google_adwords_remarketing_lists/transform.js:191]: Do not have the exact list rn but can be found in test cases input/output files. Mainly for deletion support, different scenarios of how we pass data have been added and few mentioned below are:
* breaking the user identifier array in chunks of 20
* If neither add nor remove keyword is provided then it throws a custom error.
* only passing remove operation etc.
For the `add` operation @ItsSudip has already added all the custom errors which are being handled and its scenarios.

**Review by saikumarrs — APPROVED** [review]
